### ISSUE TITLE
Object formatting

### ISSUE THREAD

**Issue by jkk4999** [CarlosNZ/json-edit-react#34]
Hi,

When adding an object, the closing parentheses are indented.  Thanks for your contribution!

<img width="350" alt="Screenshot 2024-02-25 at 9 59 43 AM" src="https://github.com/CarlosNZ/json-edit-react/assets/59376742/f8acb88c-890a-406c-b300-bf46991e5eaf">

**Comment by CarlosNZ** [comment]
Hi, that's odd. Can you provide more details of the steps you took to get this? And also, what version of the component are you using (if not the latest, try upgrading to 1.5.1)?

It looks correct if I just type that info straight into the demo system:
<img width="389" alt="Screenshot 2024-02-26 at 4 19 53 PM" src="https://github.com/CarlosNZ/json-edit-react/assets/5456533/0efd0243-8b72-412c-baeb-29a3b8e1b185">

**Comment by jkk4999** [comment]
I'm using 1.5.1. and React 18.2.0. This minimal code produces the issue.

return (
      <div >
         <JsonEditor
            data={data}
            restrictEdit={({ level }) => level === 0}
            onUpdate={({ newData }) => {
               console.log('new flowConfig data is');
               console.dir(newData)
               setData(newData)
<img width="1306" alt="Screenshot 2024-02-26 at 6 56 06 AM" src="https://github.com/CarlosNZ/json-edit-react/assets/59376742/f376b4b5-22fe-4217-aeb1-2d8af1f9b04b">

            }} />
      </div>
   )

**Comment by CarlosNZ** [comment]
Okay, I've been able to reproduce this: https://codesandbox.io/p/sandbox/wizardly-yalow-g5qlym?file=%2Fsrc%2Fstyles.css%3A3%2C18

The problem is due to a `text-align: center` CSS on an ancestor element, which is cascading down into this component. You can fix it yourself by explicitly setting `text-align: left` on the immediate parent of json-edit-react, but I'm going to do a quick fix and force this on the component itself, so you can upgrade later on and have it fixed for you :)

**Comment by CarlosNZ** [comment]
Righto, v1.5.2 out now. Please let me know if it fixes your case and I'll close the issue.

Thanks for bringing it to my attention

**Comment by jkk4999** [comment]
1.5.3 fixed the issue...much appreciated!  In my particular use case, I'm using your component to create metadata.  When adding a new top-level key, I always want the value to be of type 'object'.  Currently, it defaults to be type 'string' and then I have to change it to 'object'.   When adding child keys, I want to remove the 'object' type.  Any guidance would be appreciated.

<img width="450" alt="Screenshot 2024-02-27 at 7 31 57 AM" src="https://github.com/CarlosNZ/json-edit-react/assets/59376742/ecfbf277-a534-4515-8294-366b144a20bb">

**Comment by CarlosNZ** [comment]
> When adding a new top-level key, I always want the value to be of type 'object'. Currently, it defaults to be type 'string' and then I have to change it to 'object'.

You can set the default value prop: `defaultValue={{}}` (or `defaultValue={{myKey: "defaultValue"}}`). But this will apply to all new elements, not just the top level one, which is not what you want, right?

I guess I could make the `defaultValue` prop also take a function to determine what it should be based on data state -- would that be useful?

> When adding child keys, I want to remove the 'object' type. Any guidance would be appreciated.

You can restrict the type selection list: `restrictTypeSelection={['string', 'boolean', 'number', 'null']}`. But again, I suspect you want it to be different for the top level. So a dynamic default value is probably the answer.

Thanks for the suggestion. I've made an issue to address this: #35 

It should be fairly straightforward, I'll try and update it in the next couple of days.

**Comment by CarlosNZ** [comment]
Original issue solved

**Comment by jkk4999** [comment]
Yes, adding the ability for the default value and restrictTypeSelection props to take a function would address both issues.  Appreciate your support!

**Comment by CarlosNZ** [comment]
> Yes, adding the ability for the default value and restrictTypeSelection props to take a function would address both issues. Appreciate your support!

`restrictTypeSelection` already can take a function, so it's just the default value. I'll get on to that soon.

**Comment by CarlosNZ** [comment]
Okay, @jkk4999, you can try v1.6.0 now, with your [requested feature](#35).

You should be able to achieve what you want by specifying:
```jsx
// Create empty objects at the root level, strings elsewhere:
 defaultValue={({ level }) => {
      if (level === 0) return {}
      return 'New Value'
    }},
// Prevent changing values to objects or arrays:
restrictTypeSelection={['string', 'number', 'boolean', 'null']}
```

There's an example of something similar in the [Demo](https://carlosnz.github.io/json-edit-react/) -- if you select "Client List" data set and add a new item at the root level, you'll get a pre-filled structured object, but just plain strings elsewhere.

Thanks for the suggestion -- not sure why I hadn't done that already -- you could already specify functions for pretty much all other values. ;)

**Comment by CarlosNZ** [comment]
> Per your documentation, if you return false or a string, the internal data state won't actually be updated. However, it seems like it is

Oh, that's a good point. I should be able to change this pretty easily to only update the internal state if the update function doesn't return an error.

**Comment by CarlosNZ** [comment]
> In my use case, I need to restrict the values of child nodes to 'string', 'number' or 'boolean'. To that end, I make use of the onEdit function and inspect the value of newValue prop.

Is restricting the available types in `restrictTypeSelection` not enough?

**Comment by CarlosNZ** [comment]
> In a related issue, the error message persists for a few seconds. It would be preferable that the error message is displayed until such time as the error is resolved.

I'll have to think about this/play around with it. In light of your other comment, if we don't actually update the internal state on error, then it won't actually be in an illegal state. This is how it works for other errors, such as invalid JSON -- the internal state never gets updated, so the error message just shows for a certain time.

**Comment by CarlosNZ** [comment]
Hi @jkk4999 , just to let you know that I've published v1.6.1 which has a fix for #37. If your update function returns `false` or a string value, the internal data won't be updated and the error will show.

Hope that works for you.

**Comment by jkk4999** [comment]
Perfect.  Thanks so much for your efforts!
